### DISCUSSION TITLE
Is the docs for creating log middleware relevant?

### DISCUSSION THREAD

**Discussion by budarin** [pmndrs/zustand#2233] [Q&A] [answered]
I'm trying to put logger middleware into practice as described in the readme:
```js
const log = (config) => (set, get, api) =>
  config(
    (...args) => {
      console.log('  applying', args)
      set(...args)
      console.log('  new state', get())
    },
    get,
    api,
  )
 ```
 in code
 ```ts
 export function createAppStore(): void {
    useAppStore = create<StoreState>()(
        loggerMiddleware(() => ({
            icons: {
                byId: {},
                ids: [],
            },
        })),
    );
}
```
 
but in runtime when adding data to store I don't see logs
when debugging instead of calling log function on every change - middleware calls once on store creation time only

**Comment by dai-shi** [comment, marked as answer]
I admit the docs are a little bit confusing. I wonder how to improve docs, simply drop it, maybe.

Your `log` middleware only patches `set` to the state creator. If you need to patch the api, you would need to something like this:

```js
const log = (config) => (set, get, api) => {
  const patchedSet = (...args) => {
    console.log('  applying', args)
    set(...args)
    console.log('  new state', get())
  }
  api.setState = patchedSet
  return config(patchedSet, get, api)
}
```

So, the truth is one who wants to develop custom Zustand middleware requires a decent knowledge of Zustand implementation details. Furthermore, typing it with TypeScript is much much harder. It's best to use third-party packages, if such ones exist. (Otherwise, it's a good opportunity for Zustand middleware authors.)

**Reply by budarin** [reply]
Thanks, this code works!

In Typescript, I just mark middleware so that typescript doesn't touch it - it's easier that way 😊
